Clarify semantics of locks wrt SELECT/READ (#1614)
cap-js/docs: guides/providing-services.md

_Pessimistic locking_ allows you to lock the selected records so that other transactions are blocked from changing the records in any way.

Use _exclusive_ locks when reading entity data with the _intention to update_ it in the same transaction and you want to prevent the data to be locked or updated in a concurrent transaction.

Use _shared_ locks if you only need to prevent the entity data to be locked exclusively by an update in a concurrent transaction or by a read operation with lock mode _exclusive_. Non-locking read operations or read operations with lock mode _shared_ are not prevented.

The records are locked until the end of the transaction by commit or rollback statement.

> 

Reviewer: ```suggestion
Here's an overview table:

| State | Select Without Lock | Select With Shared Lock |  Select With Exclusive Lock/Update |
| --- | --- | --- |  --- | 
| not locked | passes -> not locked | passes -> shared lock | passes -> exclusive lock |
| shared lock | passes -> not locked | passes -> shared lock | waits -> exclusive lock |
| exclusive lock | passes -> not locked | waits -> exclusive lock | waits -> exclusive lock |

```

_Pessimistic locking_ allows you to lock the selected records so that other transactions are blocked from changing the records in any way.

Use _exclusive_ locks when reading entity data with the _intention to update_ it in the same transaction and you want to prevent the data to be locked or updated in a concurrent transaction.

> Use _shared_ locks if you only need to prevent the entity data to be updated in a concurrent transaction, but don't want to block concurrent locking read operations.

Reviewer 2: ```suggestion
Use _shared_ locks if you only need to prevent the entity data to be locked exclusively by an update in a concurrent transaction or by a read operation with lock mode _exclusive_. Non-locking read operations or read operations with lock mode _shared_ are not prevented.
```

The records are locked until the end of the transaction by commit or rollback statement.

::: info
It is allowed to read **locked** entities with a `Select` query that doesn't define a lock as well.
This applies both to _exclusive_ and _shared_ locks.
:::

[Learn more about using the `SELECT ... FOR UPDATE` statement in the Node.js runtime.](../node.js/cds-ql#forupdate){.learn-more}

[Learn more about using the `Select.lock()` method in the Java runtime.](../java/working-with-cql/query-api#write-lock){.learn-more}

> []

Reviewer 2: what's this?

_Pessimistic locking_ allows you to lock the selected records so that other transactions are blocked from changing the records in any way.

Use _exclusive_ locks when reading entity data with the _intention to update_ it in the same transaction and you want to prevent the data to be locked or updated in a concurrent transaction.

Use _shared_ locks if you only need to prevent the entity data to be locked exclusively by an update in a concurrent transaction or by a read operation with lock mode _exclusive_. Non-locking read operations or read operations with lock mode _shared_ are not prevented.

The records are locked until the end of the transaction by commit or rollback statement.

Here's an overview table:

| State              | Select Without Lock   | Select With Shared Lock |  Select With Exclusive Lock/Update |
| --------------- | ----------------------- | -------------------------- |  ------------------------------------- | 
| not locked      | passes -> not locked | passes -> shared lock    | passes -> exclusive lock                  |
| shared lock     | passes -> not locked | passes -> shared lock    | waits -> exclusive lock                     |

> | exclusive lock | passes -> not locked | waits -> exclusive lock   | waits -> exclusive lock                     |

Author: To me it is not clear what the `->` in this table means. Does it mean "and this will be the new state"? If yes I think the table is not correct. In my opinion we should remove those hints completely and only keep "passes" or "waits".
@agoerler Can you clarify?

Reviewer 2: The intent of `->` was to indicate the target state after execution of the select. I think the entry exclusive lock/select with shared lock is wrong as the select would wait until the exclusive lock is released and hence the traget state should be "shared lock". Also select without always leaves the lock state untouched. 

I agree that we should remove the `->`.

_Pessimistic locking_ allows you to lock the selected records so that other transactions are blocked from changing the records in any way.

Use _exclusive_ locks when reading entity data with the _intention to update_ it in the same transaction and you want to prevent the data to be locked or updated in a concurrent transaction.

Use _shared_ locks if you only need to prevent the entity data to be locked exclusively by an update in a concurrent transaction or by a read operation with lock mode _exclusive_. Non-locking read operations or read operations with lock mode _shared_ are not prevented.

The records are locked until the end of the transaction by commit or rollback statement.

> Here's an overview table:
> 
> | State | Select Without Lock | Select With Shared Lock |  Select With Exclusive Lock/Update |
> | --- | --- | --- |  --- | 
> | not locked | passes -> not locked | passes -> shared lock | passes -> exclusive lock |
> | shared lock | passes -> not locked | passes -> shared lock | waits -> exclusive lock |
> | exclusive lock | passes -> not locked | waits -> exclusive lock | waits -> exclusive lock |

Reviewer: ```suggestion

Here's an overview table:

| State              | Select Without Lock   | Select With Shared Lock |  Select With Exclusive Lock/Update |
| --------------- | ----------------------- | -------------------------- |  ------------------------------------- | 
| not locked      | passes -> not locked | passes -> shared lock    | passes -> exclusive lock                  |
| shared lock     | passes -> not locked | passes -> shared lock    | waits -> exclusive lock                     |
| exclusive lock | passes -> not locked | waits -> exclusive lock   | waits -> exclusive lock                     |
```

Use _exclusive_ locks when reading entity data with the _intention to update_ it in the same transaction and you want to prevent the data to be locked or updated in a concurrent transaction.

Use _shared_ locks if you only need to prevent the entity data to be locked exclusively by an update in a concurrent transaction or by a read operation with lock mode _exclusive_. Non-locking read operations or read operations with lock mode _shared_ are not prevented.

The records are locked until the end of the transaction by commit or rollback statement.

Here's an overview table:

| State              | Select Without Lock   | Select With Shared Lock |  Select With Exclusive Lock/Update |
| --------------- | ----------------------- | -------------------------- |  ------------------------------------- | 
| not locked      | passes -> not locked | passes -> shared lock    | passes -> exclusive lock                  |
| shared lock     | passes -> not locked | passes -> shared lock    | waits -> exclusive lock                     |
| exclusive lock | passes -> not locked | waits -> exclusive lock   | waits -> exclusive lock                     |

> ::: info
> It is allowed to read **locked** entities with a `Select` query that doesn't define a lock as well.
> This applies both to _exclusive_ and _shared_ locks.
> :::

Reviewer: Can this then be removed?
```suggestion
```
